### ISSUE TITLE
Order of appearance of lxc.net directives has meaning

### ISSUE THREAD

**Issue by bmoller** [lxc/lxc#3405]
# Required information

 * Distribution: Arch Linux
 * The output of
   * `lxc-start --version`: 4.0.2
   * `lxc-checkconfig`:
```
--- Namespaces ---
Namespaces: enabled
Utsname namespace: enabled
Ipc namespace: enabled
Pid namespace: enabled
User namespace: enabled
Warning: newuidmap is not setuid-root
Warning: newgidmap is not setuid-root
Network namespace: enabled

--- Control groups ---
Cgroups: enabled

Cgroup v1 mount points: 
/sys/fs/cgroup/systemd
/sys/fs/cgroup/net_cls,net_prio
/sys/fs/cgroup/perf_event
/sys/fs/cgroup/cpuset
/sys/fs/cgroup/cpu,cpuacct
/sys/fs/cgroup/rdma
/sys/fs/cgroup/freezer
/sys/fs/cgroup/devices
/sys/fs/cgroup/pids
/sys/fs/cgroup/blkio
/sys/fs/cgroup/memory
/sys/fs/cgroup/hugetlb

Cgroup v2 mount points: 
/sys/fs/cgroup/unified

Cgroup v1 clone_children flag: enabled
Cgroup device: enabled
Cgroup sched: enabled
Cgroup cpu account: enabled
Cgroup memory controller: enabled
Cgroup cpuset: enabled

--- Misc ---
Veth pair device: enabled, loaded
Macvlan: enabled, loaded
Vlan: enabled, not loaded
Bridges: enabled, loaded
Advanced netfilter: enabled, not loaded
CONFIG_NF_NAT_IPV4: missing
CONFIG_NF_NAT_IPV6: missing
CONFIG_IP_NF_TARGET_MASQUERADE: enabled, not loaded
CONFIG_IP6_NF_TARGET_MASQUERADE: enabled, not loaded
CONFIG_NETFILTER_XT_TARGET_CHECKSUM: enabled, not loaded
CONFIG_NETFILTER_XT_MATCH_COMMENT: enabled, not loaded
FUSE (for use with lxcfs): enabled, not loaded

--- Checkpoint/Restore ---
checkpoint restore: enabled
CONFIG_FHANDLE: enabled
CONFIG_EVENTFD: enabled
CONFIG_EPOLL: enabled
CONFIG_UNIX_DIAG: enabled
CONFIG_INET_DIAG: enabled
CONFIG_PACKET_DIAG: enabled
CONFIG_NETLINK_DIAG: enabled
File capabilities:
```
   * `uname -a`: Linux pogi-t610 5.4.36-1-lts #1 SMP Wed, 29 Apr 2020 16:23:03 +0000 x86_64 GNU/Linux
   * `cat /proc/self/cgroup`:
```
12:hugetlb:/
11:memory:/user.slice/user-1000.slice/session-6.scope
10:blkio:/user.slice
9:pids:/user.slice/user-1000.slice/session-6.scope
8:devices:/user.slice
7:freezer:/
6:rdma:/
5:cpu,cpuacct:/user.slice
4:cpuset:/
3:perf_event:/
2:net_cls,net_prio:/
1:name=systemd:/user.slice/user-1000.slice/session-6.scope
0::/user.slice/user-1000.slice/session-6.scope
```
   * `cat /proc/1/mounts`:
```
proc /proc proc rw,nosuid,nodev,noexec,relatime 0 0
sys /sys sysfs rw,nosuid,nodev,noexec,relatime 0 0
dev /dev devtmpfs rw,nosuid,relatime,size=49465520k,nr_inodes=12366380,mode=755 0 0
run /run tmpfs rw,nosuid,nodev,relatime,mode=755 0 0
efivarfs /sys/firmware/efi/efivars efivarfs rw,nosuid,nodev,noexec,relatime 0 0
pogi-pool/arch/root / zfs rw,relatime,xattr,posixacl 0 0
pogi-pool/arch/var /var zfs rw,relatime,xattr,posixacl 0 0
pogi-pool/lxc/alpine /var/lib/lxc/alpine/rootfs zfs rw,relatime,xattr,posixacl 0 0
pogi-pool/lxc/emby /var/lib/lxc/emby/rootfs zfs rw,relatime,xattr,posixacl 0 0
pogi-pool/lxc/nginx /var/lib/lxc/nginx/rootfs zfs rw,relatime,xattr,posixacl 0 0
pogi-pool/lxc/unifi /var/lib/lxc/unifi/rootfs zfs rw,relatime,xattr,posixacl 0 0
pogi-pool/samba/brandon /srv/samba/brandon zfs rw,relatime,xattr,posixacl 0 0
pogi-pool/samba/games /srv/samba/games zfs rw,relatime,xattr,posixacl 0 0
pogi-pool/samba/kerri /srv/samba/kerri zfs rw,relatime,xattr,posixacl 0 0
pogi-pool/samba/music /srv/samba/music zfs rw,relatime,xattr,posixacl 0 0
pogi-pool/samba/pictures /srv/samba/pictures zfs rw,relatime,xattr,posixacl 0 0
pogi-pool/samba/video /srv/samba/video zfs rw,relatime,xattr,posixacl 0 0
securityfs /sys/kernel/security securityfs rw,nosuid,nodev,noexec,relatime 0 0
tmpfs /dev/shm tmpfs rw,nosuid,nodev 0 0
devpts /dev/pts devpts rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000 0 0
tmpfs /sys/fs/cgroup tmpfs ro,nosuid,nodev,noexec,mode=755 0 0
cgroup2 /sys/fs/cgroup/unified cgroup2 rw,nosuid,nodev,noexec,relatime,nsdelegate 0 0
cgroup /sys/fs/cgroup/systemd cgroup rw,nosuid,nodev,noexec,relatime,xattr,name=systemd 0 0
pstore /sys/fs/pstore pstore rw,nosuid,nodev,noexec,relatime 0 0
none /sys/fs/bpf bpf rw,nosuid,nodev,noexec,relatime,mode=700 0 0
cgroup /sys/fs/cgroup/net_cls,net_prio cgroup rw,nosuid,nodev,noexec,relatime,net_cls,net_prio 0 0
cgroup /sys/fs/cgroup/perf_event cgroup rw,nosuid,nodev,noexec,relatime,perf_event 0 0
cgroup /sys/fs/cgroup/cpuset cgroup rw,nosuid,nodev,noexec,relatime,cpuset 0 0
cgroup /sys/fs/cgroup/cpu,cpuacct cgroup rw,nosuid,nodev,noexec,relatime,cpu,cpuacct 0 0
cgroup /sys/fs/cgroup/rdma cgroup rw,nosuid,nodev,noexec,relatime,rdma 0 0
cgroup /sys/fs/cgroup/freezer cgroup rw,nosuid,nodev,noexec,relatime,freezer 0 0
cgroup /sys/fs/cgroup/devices cgroup rw,nosuid,nodev,noexec,relatime,devices 0 0
cgroup /sys/fs/cgroup/pids cgroup rw,nosuid,nodev,noexec,relatime,pids 0 0
cgroup /sys/fs/cgroup/blkio cgroup rw,nosuid,nodev,noexec,relatime,blkio 0 0
cgroup /sys/fs/cgroup/memory cgroup rw,nosuid,nodev,noexec,relatime,memory 0 0
cgroup /sys/fs/cgroup/hugetlb cgroup rw,nosuid,nodev,noexec,relatime,hugetlb 0 0
systemd-1 /proc/sys/fs/binfmt_misc autofs rw,relatime,fd=29,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=29584 0 0
mqueue /dev/mqueue mqueue rw,nosuid,nodev,noexec,relatime 0 0
debugfs /sys/kernel/debug debugfs rw,nosuid,nodev,noexec,relatime 0 0
hugetlbfs /dev/hugepages hugetlbfs rw,relatime,pagesize=2M 0 0
tracefs /sys/kernel/tracing tracefs rw,nosuid,nodev,noexec,relatime 0 0
tmpfs /tmp tmpfs rw,nosuid,nodev 0 0
configfs /sys/kernel/config configfs rw,nosuid,nodev,noexec,relatime 0 0
/dev/sdg1 /efi vfat rw,nosuid,nodev,noexec,relatime,fmask=0022,dmask=0022,codepage=437,iocharset=iso8859-1,shortname=mixed,utf8,errors=remount-ro 0 0
/dev/sdg1 /boot vfat rw,nosuid,nodev,noexec,relatime,fmask=0022,dmask=0022,codepage=437,iocharset=iso8859-1,shortname=mixed,utf8,errors=remount-ro 0 0
pogi-pool/arch/home /home zfs rw,relatime,xattr,posixacl 0 0
binfmt_misc /proc/sys/fs/binfmt_misc binfmt_misc rw,nosuid,nodev,noexec,relatime 0 0
tmpfs /run/user/1000 tmpfs rw,nosuid,nodev,relatime,size=9897584k,mode=700,uid=1000,gid=1000 0 0
```

# Issue description

**TLDR**: at least some `lxc.container.conf` directives are positional, which is not documented anywhere. This leads to unexpected behavior and network symptoms that are nearly impossible to diagnose.

### In-Depth

Setting the mode of a `macvlan` network interface before its type causes the mode to be reset and revert to its default behavior. I use LXC completely independently of LXD and compiled a multi-container configuration that allowed containers to communicate over a `macvlan`-based network. This was written following the arguments in order of presentation from the `lxc.container.conf` manpage.

Once I had my initial implementation working I sorted the directives alphabetically out of habit and in order to find things easily. After rebooting the host the next morning, containers could no longer ping each other or connect in any way. I spent literally the entire day trying to figure out what was wrong with little success.

Out of desperation I finally tried using `ipvlan` interfaces instead. Changing only the `type` and `mode` arguments gave me this `/etc/lxc/default` which I include in my container configs:

```
lxc.net.0.flags = up
lxc.net.0.ipv4.gateway = 192.168.1.1
lxc.net.0.link = br0
lxc.net.0.ipvlan.isolation = bridge
lxc.net.0.ipvlan.mode = l3
lxc.net.0.type = ipvlan
```
On `lxc-start` this error is returned:

```
sudo lxc-start nginx
lxc-start: nginx: confile.c: set_config_net_ipvlan_isolation: 525 Success - Invalid ipvlan isolation "bridge", can only be used with ipvlan network
lxc-start: nginx: parse.c: lxc_file_for_each_line_mmap: 120 Failed to parse config file "/etc/lxc/default.conf" at line "lxc.net.0.ipvlan.isolation = bridge"
lxc-start: nginx: parse.c: lxc_file_for_each_line_mmap: 120 Failed to parse config file "/var/lib/lxc/nginx/config" at line "lxc.include = /etc/lxc/default.conf"
Failed to load config for nginx
lxc-start: nginx: tools/lxc_start.c: main: 242 Failed to create lxc_container
```

This was a clue to me that position of these directives matters, and indeed this works as expected:

```
lxc.net.0.type = ipvlan
lxc.net.0.flags = up
lxc.net.0.ipv4.gateway = 192.168.1.1
lxc.net.0.link = br0
lxc.net.0.ipvlan.isolation = bridge
lxc.net.0.ipvlan.mode = l3
```

Similarly, this configuration I believe causes the `macvlan` interfaces to revert to default `private` behavior, as I get `No route to host` between containers:

```
lxc.net.0.flags = up
lxc.net.0.ipv4.gateway = 192.168.1.1
lxc.net.0.link = br0
lxc.net.0.macvlan.mode = bridge
lxc.net.0.type = macvlan
```

This configuration however works as expected:

```
lxc.net.0.type = macvlan
lxc.net.0.flags = up
lxc.net.0.ipv4.gateway = 192.168.1.1
lxc.net.0.link = br0
lxc.net.0.macvlan.mode = bridge
```

In other words, the `lxc.net.[i].type` must appear before certain other directives that affect that interface's operation. I'm guessing that this is a holdover of legacy versions of LXC which used the `type` directive to begin a new block of interface configuration.

In my mind order shouldn't matter in these configuration files, with the possible exception of `lxc.include` statements. At the very least the importance of order should be documented in the `lxc.container.conf` manpage.

For the moment, all appears to function as intended if I order per appearance in the manpage and set aside my OCD ;)

# Steps to reproduce

 1. Place `lxc.net.[i].macvlan.mode = bridge` before the same interface's `lxc.net.[i].type`
 2. Observe that containers appear to operate in `private` mode
 3. Place the `lxc.net.[i].type` before all other directives for that interface
 4. Observe that containers are now able to connect to each other as expected for this mode

**Comment by hallyn** [comment]
You say:

```
at least some lxc.container.conf directives are positional, which is not documented anywhere
```

The lxc.container.conf manpage says:

```
       Linux  containers are defined with a simple configuration file. Each option in the configuration file has the
       form key = value fitting in one line. The "#" character means the line is a comment. List options, like capa‐
       bilities  and  cgroups  options, can be used with no value to clear any previously defined values of that op‐
       tion.
```

Clearly since an empty list option can "clear any previously defined values", position matters.  It has always been so.  This is crucial to be able to override pieces inherited from 'include'd files.

If you have a suggestion for making this clearer in the manpage, a PR is most definately welcome.  It's definately not meant to be confusing or frustrating.

**Comment by bmoller** [comment]
I think that's taking that specific line I wrote a little out of context. If you read the whole thing the specific behavior is in reference to the `lxc.net.*` directives which are in fact not documented as being positional. The one exception is `lxc.net` without any arguments, which is stated as resetting all previous networking configuration.

My specific example shows that defining `lxc.net.[i].macvlan.mode` before `lxc.net.[i].type` will cause LXC to discard the mode _while retaining the other configuration directives_. This is not internally consistent behavior, which makes me lean toward a bug classification, not documentation.

This specific instance could be documented, but the fact that this behavior exists at all suggests that maybe a look at the logic is in order to determine if other similar conditions are lurking beneath the surface.

**Comment by hallyn** [comment]
Gotcha, thanks.

**Comment by brauner** [comment]
@bmoller can you try https://github.com/lxc/lxc/pull/3408 whether that fixes it?
